R/dt_spanners.R

```diff
@@ -137,106 +113,80 @@ dt_spanners_print_matrix <- function(
 
   spanners_tbl <- dt_spanners_get(data = data)
 
-  remove_entry <- rep(FALSE, nrow(spanners_tbl))
-
   if (!include_hidden) {
     vars <- dt_boxhead_get_vars_default(data = data)
   } else {
     vars <- dt_boxhead_get_vars(data = data)
   }
 
+  # If `spanners_tbl` is immediately empty then return a single-row
+  # matrix of column vars/IDs (or not, if `omit_columns_row = TRUE`)
+  if (nrow(spanners_tbl) < 1) {
+    return(
+      empty_spanner_matrix(vars = vars, omit_columns_row = omit_columns_row)
+    )
+  }
+
   # Keep only the columns in `spanners_tbl` that are in the `vars`;
   # this will exclude spanners if their columns aren't in `vars` (this
   # is possible when hiding columns via `col_hide()`
-  for (i in seq_len(nrow(spanners_tbl))) {
-
-    vars_spanner_i <- unlist(spanners_tbl[i, ][["vars"]])
-    vars_spanner_i_flt <- base::intersect(vars_spanner_i, vars)
-
-    if (length(vars_spanner_i_flt) < 1) {
-      # If none of the columns under the proposed spanner are
-      # set as visible columns, then we need to mark the spanner
-      # entry for deletion from `spanners_tbl`
-      remove_entry[i] <- TRUE
-    } else {
-      spanners_tbl[i, ][["vars"]] <- list(vars_spanner_i_flt)
-    }
-  }
-
-  # Remove entries from `spanners_tbl` as needed
-  spanners_tbl <- spanners_tbl[!remove_entry, ]
+  spanners_tbl <-
+    spanners_tbl %>%
+    dplyr::mutate(vars = lapply(.data$vars, base::intersect, .env$vars)) %>%
+    dplyr::filter(vapply(vars, length, integer(1)) > 0) %>%
+    dplyr::mutate(
+      spanner_level = match(spanner_level, sort(unique(spanner_level)))
+    )
 
-  # Get the maximum height of the spanner matrix; this won't
-  # necessarily be the finalized height since we will collapse
-  # empty rows
+  # If `spanners_tbl` is immediately empty then return a single-row
+  # matrix of column vars/IDs (or not, if `omit_columns_row = TRUE`)
   if (nrow(spanners_tbl) < 1) {
-    spanner_height <- 0
-  } else {
-    spanner_height <- max(spanners_tbl[["spanner_level"]])
+    return(
+      empty_spanner_matrix(vars = vars, omit_columns_row = omit_columns_row)
+    )
   }
 
+  # Get the height of the spanner matrix
+  spanner_height <- max(spanners_tbl[["spanner_level"]])
+
   vars_vec <- rep(NA_character_, length(vars))
   names(vars_vec) <- vars
 
-  # Initialize matrix to serve as boxhead representation
-  columns_mat <- matrix(vars, nrow = 1, ncol = length(vars))
+  # Initialize matrix row that represents the column vars
+  columns_mat_cols_row <- matrix(vars, nrow = 1, ncol = length(vars))
```

Comment:
Consider moving this down to right before the one time it's used

Reply:
Now moved.